### ISSUE TITLE
Read YANG library as described in RFC8525

### ISSUE THREAD

**Issue by gusmb** [CZ-NIC/yangson#126]
Current version of Yangson initializes the data model from a YANG library file from [RFC7895](https://www.rfc-editor.org/rfc/rfc7895). This is deprecated by [RFC8525](https://www.rfc-editor.org/rfc/rfc8525)

New YANG library tree:

```
module: ietf-yang-library
     +--ro yang-library
        +--ro module-set* [name]
        |  +--ro name                  string
        |  +--ro module* [name]
        |  |  +--ro name         yang:yang-identifier
        |  |  +--ro revision?    revision-identifier
        |  |  +--ro namespace    inet:uri
        |  |  +--ro location*    inet:uri
        |  |  +--ro submodule* [name]
        |  |  |  +--ro name        yang:yang-identifier
        |  |  |  +--ro revision?   revision-identifier
        |  |  |  +--ro location*   inet:uri
        |  |  +--ro feature*     yang:yang-identifier
        |  |  +--ro deviation*   -> ../../module/name
        |  +--ro import-only-module* [name revision]
        |     +--ro name         yang:yang-identifier
        |     +--ro revision     union
        |     +--ro namespace    inet:uri
        |     +--ro location*    inet:uri
        |     +--ro submodule* [name]
        |        +--ro name        yang:yang-identifier
        |        +--ro revision?   revision-identifier
        |        +--ro location*   inet:uri
        +--ro schema* [name]
        |  +--ro name          string
        |  +--ro module-set*   -> ../../module-set/name
        +--ro datastore* [name]
        |  +--ro name      ds:datastore-ref
        |  +--ro schema    -> ../../schema/name
        +--ro content-id    string
```

An idea would be to add support for both formats on the SchemaData constructor, defaulting to the new one.

**Comment by llhotka** [comment]
I have already thought about this. However, the format of RFC 7895 is used in Yangson as a convenient way of specifying the data model, and it perfectly suffices for this purpose. RFC 8525 introduces datastores that Yangson doesn't implement, so it is just an extra complication. So we'd need to decide which datastore will be used.

The only use case I can see is that one can reuse YANG Library data that may already be available.

**Comment by gusmb** [comment]
The location field could be useful as well. Right now it is assumed that all modules are locally available, but it could be improved to retrieve remote modules from a url location during data model initialization...

**Comment by gusmb** [comment]
I am using this idea for a CI/CD use case, where I load   instance data from a distributed source of truth, and export the assembled instance once as a pipeline artifact, which is then used as inventory source for the rest of the process. The ability to download shared modules comes handy here as allows to have shared modules with data types in a centralized place, to be shared among different service models. I only need to do this once when loading the instance.
So to get around the Library issue I am using the current  RFC7895 format adding the location field, which I then use to download the module through am external process. However, I think it would be nice to be able to leverage the standard RFC format and be able to use this library natively for that.

**Comment by llhotka** [comment]
Would it be sufficient for you to have a stand-alone tool accepting RFC 8525 format and:
1. convert it to the 7895 format, and
2. (optionally) download all modules?

**Comment by gusmb** [comment]
> Would it be sufficient for you to have a stand-alone tool accepting RFC 8525 format and:
> 
> 1. convert it to the 7895 format, and
> 2. (optionally) download all modules?

Yes this would work. Modules for which a remote location is specified would be downloaded if desired and yangson would find them locally when loading the library

**Comment by llhotka** [comment]
The conversion from 8525 to 7895 format can be done by the script `convert8525`, see [man page](https://yangson.labs.nic.cz/convert8525-man.html). It is included in version 1.5.0 (9d2140065ea424d8ec1c7ba85884d664859560ee).

Downloading the YANG modules is unrelated to this and can in fact be used with the old 7895 format, too. I am thinking about including it as an option in the `yangson` script.
